### ISSUE TITLE
Insert of Array via named parameter not working

### ISSUE THREAD

**Issue by rbeeli** [DarkWanderer/ClickHouse.Client#41]
Hi @DarkWanderer,

I'm trying to insert an array of float values `float[]` via a named parameter, but it gives me a broken pipe exception:
```
Exception has occurred: CLR/System.Net.Http.HttpRequestException
Exception thrown: 'System.Net.Http.HttpRequestException' in System.Private.CoreLib.dll: 'An error occurred while sending the request.'
 Inner exceptions found, see $exception in variables window for more details.
 Innermost exception 	 System.Net.Sockets.SocketException : Broken pipe
```
I'm using version `2.0.0.262`.

If I use the syntax `[1.2342,0.3234,2234.1]` in the SQL INSERT query directly, it works. If I use the named parameter syntax `{{values:Array(Float32)}}`, it doesn't. Am I using the wrong syntax or is there a bug in the Client library?

Thanks!

**Comment by DarkWanderer** [comment]
Hi,

Insertion of arrays definitely works - it's one of test cases, so there's most likely an issue with either syntax or HttpClient. Could you please share a bit more details?

1. What is the resulting SQL query (full string)?
2. What is the server version?
3. What is the inner exception?

I suspect that the syntax is not correct here (`{{values:Array(Float32)}}`), but I'm not sure how does `Broken pipe` come into equation.

Here is example code which works for me
```
using ClickHouse.Client.ADO;
using ClickHouse.Client.Utility;

            using var connection = new ClickHouseConnection(connectionString);
            await connection.OpenAsync();
            await connection.ExecuteStatementAsync("CREATE DATABASE IF NOT EXISTS test");
            await connection.ExecuteStatementAsync("TRUNCATE TABLE IF EXISTS test.float_array");
            await connection.ExecuteStatementAsync("CREATE TABLE IF NOT EXISTS test.float_array (arr Array(Float64)) ENGINE Memory");

            var command = connection.CreateCommand();
            command.AddParameter("values", new[] { 1.0, 2.0, 3.0 });
            command.CommandText = "INSERT INTO test.float_array VALUES ({values:Array(Float32)})";
            await command.ExecuteNonQueryAsync();

            var count = await connection.ExecuteScalarAsync("SELECT COUNT(*) FROM test.float_array");
```
Note that the parameter expression is wrapped in curly braces, _then_ in round brackets: `(` `{` `name`:`type` `}` `)`

**Comment by rbeeli** [comment]
Hi @DarkWanderer,

It works with the round/curly braces combination if I use a few values `({values:Array(Float32)})`. Can you try it on your end with more values, i.e. `values = Enumerable.Range(0, 10000).Select(e => (float)e).ToArray();`? It works with a few, but not with many, e.g. thousands.

New error message:
 ```
Exception thrown: 'System.UriFormatException' in System.Private.CoreLib.dll: 'Invalid URI: The Uri string is too long.
 ```
I suppose the request is sent via POST, so the URL really shouldn't get that long? It works if I only send 10 values, but not with 10000.

1. What is the resulting SQL query (full string)?
 ```
                         INSERT INTO test.arr (
                              dt_from,
                              dt_to,
                              values
                          )
                          VALUES (
                              toDateTime64({dt_from:String}, 9),
                              toDateTime64({dt_to:String}, 9),
                              ({values:Array(Float32)})
                          )
```

2. What is the server version?
`ClickHouse server version 20.8.2 revision 54438`

3. What is the runtime you use? (CoreCLR 2.0?)
`netcoreapp3.1`

4. Do you use a proxy when connecting?
`No, all localhost with default settings on Ubuntu 18 based distribution.`

Thanks for the quick response!

**Comment by DarkWanderer** [comment]
Hi,

Due to how HTTP interface works, there's only so many parameters you can push into a query before URL becomes too long. Unfortunately there is currently no way to pass query params as POST params (or at least I don't know of one)

However, for inserting large volumes, there is a solution - [bulk insertion](https://github.com/DarkWanderer/ClickHouse.Client/wiki/Bulk-insertion) functionality. You can insert (almost) unlimited number of records using code similar to following:

```
using var connection = new ClickHouseConnection(connectionString);

using var bulkCopyInterface = new ClickHouseBulkCopy(connection)
{
    DestinationTableName = "<database>.<table>",
    BatchSize = 100000
};

// Example data to test
var values = Enumerable.Range(0, count).Select(i => new object[] { (long)i });
await bulkCopyInterface.WriteToServerAsync(values);
Console.WriteLine(bulkCopyInterface.RowsWritten);
```
Basically, you need to pack each row into an object[] and pass an `IEnumerable<object[]>` to the class, it will do the rest for you. Inserting arrays like in your case is also supported

**Comment by rbeeli** [comment]
Hi,
The `ClickHouseBulkCopy` solution works, thanks, I will go with that. I guess that this is also a faster solution compared to `INSERT`-queries. Passing the query as POST body sounds something reasonable to provide for an HTTP API, or at least the parameters thereof, but I'm not familiar with the ClickHouse implementation.
Cheers!

**Comment by DarkWanderer** [comment]
That's the problem with ClickHouse - to my knowledge, you can either pass SQL in POST body and parameters in URL, or pass SQL in URL and parameters in body. I use first approach for regular queries (because they are potentially very large `SELECT`s), and second approach for `BulkCopy` (because I know the query is small and there is a lot of parameters). If I figure out a more universal approach, I'll implement it in one of next versions.

Glad the proposed solution works for you
